File: QytetetRuby/lib/especulador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module ModeloQytetet
  class Especulador < Jugador
    attr_reader :fianza
    def initialize(jugador, fianza)
      @encarcelado = jugador.encarcelado
      @nombre = jugador.nombre
      @saldo = jugador.saldo
      @casilla_actual = jugador.casilla_actual
      @carta_libertad = jugador.carta_libertad
      @carta_loteria = jugador.carta_loteria
      @propiedades = jugador.propiedades
      @fianza = fianza
    end
    
    def convertirme(_)
      self 
    end
    
    def pagar_impuestos(cantidad)
      super(cantidad/2)
    end
    
    def ir_a_carcel(casilla)
      super if !pagar_fianza(@fianza)
    end
    
    def pagar_fianza(cantidad)
      pagar_libertad(cantidad)
    end
    
    def to_s
      super + " (Especulador)"
    end
    
    def factor_especulador
      2
    end
    
  end
end

File: QytetetJava/src/modeloqytetet/Especulador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

/**
 *
 * @author fran
 */
public class Especulador extends Jugador {

    private int fianza;

    protected Especulador(Jugador jugador, int fianza) {
        super(jugador);
        this.fianza = fianza;
        factorEspeculador = 2;
    }

    @Override
    protected void pagarImpuestos(int cantidad) {
        super.pagarImpuestos(cantidad/2);
    }

    @Override
    protected void irACarcel(Casilla casilla) {
        if(!pagarFianza(fianza))
            super.irACarcel(casilla);
    }

    @Override
    protected Especulador convertirme(int fianza) {
        return this;
    }

    private boolean pagarFianza(int cantidad) {
                return pagarLibertad(cantidad);
    }
    
    @Override
    public String toString(){
        return super.toString() + " (Especulador)";
    }
    
    @Override
    public int getFactorEspeculador(){
        return factorEspeculador;
    }
}
